README.md
<div align="center">
  <h1 align="center">BombusCV</h1>

  ![GitHub releases](https://img.shields.io/github/downloads/marcoradocchia/bombuscv-rs/total?color=%23a9b665&logo=github)
  ![GitHub source size](https://img.shields.io/github/languages/code-size/marcoradocchia/bombuscv-rs?color=ea6962&logo=github)
  ![GitHub open issues](https://img.shields.io/github/issues-raw/marcoradocchia/bombuscv-rs?color=%23d8a657&logo=github)
  ![GitHub open pull requests](https://img.shields.io/github/issues-pr-raw/marcoradocchia/bombuscv-rs?color=%2389b482&logo=github)
  ![GitHub sponsors](https://img.shields.io/github/sponsors/marcoradocchia?color=%23d3869b&logo=github)
  ![Crates.io downloads](https://img.shields.io/crates/d/bombuscv-rs?label=crates.io%20downloads&color=%23a9b665&logo=rust)
  ![Crates.io version](https://img.shields.io/crates/v/bombuscv-rs?logo=rust&color=%23d8a657)
  ![Discord](https://img.shields.io/discord/985154521946816595?label=chat%20support&logo=discord&logoColor=%23ffff&color=%2389b482)
  ![GitHub license](https://img.shields.io/github/license/marcoradocchia/bombuscv-rs?color=%23e78a4e)
</div>

Motion detection & video recording software based on OpenCV, built for research
on **Bumblebees** (hence the name).

## Index

- [Use case](#use-case)
- [Examples](#examples)
- [Install](#install)
  - [Requirements](#requirements)
  - [Cargo](#cargo)
  - [Install on RaspberryPi 4](#install-on-raspberrypi-4)
- [Usage](#usage)
- [Configuration](#configuration)
- [Changelog](#changelog)
- [ToDo](#todo)
- [Chat Support](#chat-support)
- [License](#license)

## Use case

This software was built to meet the need of tracking, and/or recording clips of
marked Bumblebee individuals in a scientific research project. It has been used
with a
[Raspberry Pi 4](https://www.raspberrypi.com/products/raspberry-pi-4-model-b/)[^1]
and a
[Raspberry Pi HQ Camera](https://www.raspberrypi.com/products/raspberry-pi-high-quality-camera/)[^2]
pointed at the entrance of a _Bombus terrestris_ nest, in order to record clips
of the entry/exit events, based on motion. This considerably reduced the
storage space required for the recordings and completely removed the need of
post processing work, since it was only recording clips in which individuals
appeared in the video frame.

`bombuscv-rs` offers realtime motion detection & video recording[^3] using
camera input and can be directly used on fieldwork. However, using the `video`
option, live camera input can be replaced with a pre-recorded video file: this
is useful to _remove dead moments_ from videos and reduce/remove the need of
manual video trimming.

[^1]: 4GB of RAM memory, powered by a 30000mAh battery power supply, which
  means this setup can be also reproduced in locations where no AC is available
[^2]: 12.3 megapixel _Sony IMX477_ sensor
[^3]: Based on hardware (RasberryPi 4 at the moment of writing can handle
  640x480 resolution at 60fps)

## Examples

Below a brief example of the produced video output:

<div align="center">

https://user-images.githubusercontent.com/74802223/171311278-c5caf303-832f-46f6-a4cc-a3e05f823349.mp4

</div>

More examples can be found at the [BombusCV YouTube
channel](https://www.youtube.com/channel/UCkSjz-EAjhEvtUcY-ruYkrA).

## Install

For installation on *RaspberryPi* check [Install on RaspberryPi
4](#install-on-raspberrypi-4).

### Requirements

This program requires a working installation of **OpenCV** (`>=4.5.5`).
Building OpenCV from source is recommended (if you're going to build OpenCV
from source make sure to also install OpenCV dependencies), although it should
work with precompiled packages in your distro's repositories (it has been
tested with success on *ArchLinux* with the `extra/opencv` package).

### Cargo

A package is available at [crates.io](https://crates.io/crates/bombuscv-rs). In
order to install it run `cargo install bombuscv-rs` in your shell[^4].

[^4]: Assuming Rust installed

### Install on RaspberryPi 4

It is strongly recommended to use a RaspberryPi 4 with at least 4GB of RAM.
Also, before trying to install, please enable *Legacy Camera* support under
*Interface options*  running `raspi-config` and reboot. Since installation on a
RaspberryPi may be a little bit *tricky*, an installation script is
provided[^5]. It takes care of updating & preparing the system, compiling
*OpenCV* and installing *Rustup* and finally **BombusCV**. You can run the
[instllation script](bombuscv-raspi.sh) using `curl`:
```sh
curl --proto '=https' --tlsv1.2 -sSf https://raw.githubusercontent.com/marcoradocchia/bombuscv-rs/master/bombuscv-raspi.sh | sh
```

[^5]: RaspberryPi OS 64 bits required in order to install using the script

## Usage

```
bombuscv-rs 0.4.0
Marco Radocchia <marco.radocchia@outlook.com>
OpenCV based motion detection/recording software built for research on bumblebees.

USAGE:
    bombuscv [OPTIONS]

OPTIONS:
    -c, --codec <CODEC>            Video codec [possible values: h264, mjpg, xvid, mp4v]
    -d, --directory <DIRECTORY>    Output video directory
    -f, --framerate <FRAMERATE>    Video capture framerate
        --format <FORMAT>          Output video filename format (see
                                   <https://docs.rs/chrono/latest/chrono/format/strftime/index.html>
                                   for valid specifiers)
    -h, --help                     Print help information
    -H, --height <HEIGHT>          Video capture frame height
    -i, --index <INDEX>            /dev/video<INDEX> capture camera index
        --no-color                 Disable colored output
    -o, --overlay                  Date&Time video overlay
    -q, --quiet                    Mute standard output
    -v, --video <VIDEO>            Video file as input
    -V, --version                  Print version information
    -W, --width <WIDTH>            Video capture frame width
```

Specifying `width`, `height` & `framerate` will make `bombuscv` probe the
capture device for the closest combination of values it can provide and select
them. In other words: if you required valid options, they will be used,
otherwhise `bombuscv` will adapt those to the closest available combination[^6].

Note that `video` option, which runs `bombuscv` with a pre-recorded video
input, is incompatible with `framerate`, `width`, `height` and `overlay`. Also,
if these options are specified in the configuration file, they are going to be
ignored. This because the first two are auto-detected from the input file while
the last makes no sense if used with a non-live video feed; same rules apply to
CLI arguments.

[^6]: Same rules apply to configuration file

## Configuration

All CLI options (except `video` and `no-color`) can be set in a *optional* configuration file
stored at `$XDG_CONFIG_HOME/bombuscv/config.toml` by default or at any other
location in the filesystem specified by setting `BOMBUSCV_CONFIG` environment
variable. CLI options/arguments/flags override those defined in the
configuration file. Below listed an example configuration file:
```toml
# be quiet (mute stdout)
quiet = false
# output video directory
directory = "~/output_directory/"
# output video filename format (see
# https://docs.rs/chrono/latest/chrono/format/strftime/index.html for valid specifiers)
format = "%Y-%m-%dT%H:%M:%S"

# The following options are ignored if bombuscv is run with `--video` option
# /dev/video<index> camera input
index = 0
# video capture frame width
width = 640
# video capture frame height
height = 480
# video capture framerate
framerate = 30
# video codec [possible values: "h264", "mjpg", "xvid", "mp4v"]
codec = "h264"
# date&time video overlay
overlay = true
# date&time video overlay border
overlay_border = 2
```

## Changelog

Complete [CHANGELOG](CHANGELOG.md).

## ToDo

- [x] Provide build & install instructions in [README](README.md), as well as
  the instructions to install OpenCV.
- [x] Make install script for automated installation on RaspberryPi.
- [x] Passing `video` or `directory` options in the configuration file using
  `~/<path>` results in an error: in the Deserialize expanding `~` to
  absolute path is required.
- [x] Using `video`, _date&time_ overlay generated on frame grabbed makes no
  sense: disable video overlay while using `video` option.
- [x] Add option to specify custom config path using env variables.
- [x] Add option to specify (in config file or via CLI argument) a custom
  output video filename formatter (must be [chrono DateTime
  syntax](https://docs.rs/chrono/latest/chrono/format/strftime/index.html)).
- [x] Add thread signalling to interrupt grabber thread and gracefully
  terminate the execution.
- [x] Move logic from `main` to newly defined `run`.

## Chat Support

Join *BombusCV's Discord server* for installation or usage chat support:

<div align="center">

[![Join our Discord server!](https://invidget.switchblade.xyz/srNGQEs2QA?language=en)](http://discord.gg/srNGQEs2QA)

</div>

## License

[GPLv3](LICENSE)


CHANGELOG.md
# Changelog

## [Unreleased]

Updated 
[README](https://github.com/marcoradocchia/bombuscv-rs/blob/master/README.md)
*Configuration* and *Usage* section.

### Added

- Configuration option to specify *Date&Time overlay border*.
- Configuration and CLI options to specify *video codec*.

### Changed

- Create `directory` (as specified in CLI option or configuration file) if it
  doesn't exist, rather than using default configuration.

## [0.3.0] - 2022-06-27

Updated
[README](https://github.com/marcoradocchia/bombuscv-rs/blob/master/README.md)
*Use case*, *Install*, *Usage* and *Configuration* sections.

### Added

- Colored output with `no-color` CLI option to disable it.

### Fixed

- Bug in motion detection preventing it from working as expected.

### Changed

- Moved `resolution` option to `width` & `height` options in both configuration
  file and CLI arguments: now custom resolution (and aspect ratio) and
  framerate can be specified and `bobmuscv` will adapt those to the closest
  combination of resolution and framerate the capture device provides.

### Removed

- Dependency `ffprobe`: pre-recorded video *resolution* and *framerate*,
  required to construct the `VideoWriter`, are now obtained using **OpenCV**'s
  `VideoCapture` getters methods.

## [0.2.0] - 2022-06-14

Updated
[README](https://github.com/marcoradocchia/bombuscv-rs/blob/master/README.md)
*Examples*, *Usage* and *Configuration* sections.

### Changed

- Updated `bombuscv-raspi.sh` script to install **OpenCV** `v4.6.0`.

### Fixed

- Issue marcoradocchia/bombuscv-rs#1 which prevented `bombuscv` to autodetect
  video framerate and resolution on `video` CLI option used.

### Removed

- Option to specify `video` in the configuration file in favor of passing video
  file via CLI argument.
- Dependency `validator`: moved config file options validation to `serde`.

## [0.1.1] - 2022-06-11

Updated
[README](https://github.com/marcoradocchia/bombuscv-rs/blob/master/README.md)
*Install* section.

### Added

- [bombuscv-raspi.sh](https://github.com/marcoradocchia/bombuscv-rs/blob/master/bombuscv-raspi.sh)
  for automated build & installation on RaspberryPi (RaspberryPi OS AArch64).

### Fixed

- Issue marcoradocchia/bombuscv-rs#2 which caused compilation errors on AArch64 systems.

## [0.1.0] - 2022-06-06

- Initial release.


src/test.rs
use crate::{Codec, Config, Grabber, Local, MotionDetector, Path, Writer};
use bombuscv_rs::Frame;
use directories::BaseDirs;
use std::{fs, time::Instant};

#[test]
fn sync_frame_processing_avg_time() {
    // Number of frames to acquire.
    const N: usize = 2000;

    // Generate the absolute path for HOME.
    let home = BaseDirs::new().unwrap().home_dir().to_path_buf();

    // Parse CLI arguments.
    let config = Config {
        index: 0,
        height: 1080,
        width: 1920,
        framerate: 60,
        video: Some(home.join("test.mkv")),
        directory: home,
        format: String::from("output"),
        overlay: false,
        no_color: true,
        quiet: false,
    };

    // Format video file path as <config.directory/date&time>.
    let filename = Local::now()
        .format(
            config
                // Output video file directory.
                .directory
                // Output video file name (derived by file format) + extension.
                .join(Path::new(&config.format).with_extension("mkv"))
                // Convert Path object to string;
                .to_str()
                .unwrap(),
        )
        .to_string();

    // Vector of frames to test performance on.
    let mut frames: Vec<Frame> = Vec::with_capacity(N);
    let mut detected_frames = 0;

    // Instance of the frame grabber.
    let mut grabber = match &config.video {
        // VideoCapture is video file.
        Some(video) => Grabber::from_file(video),
        // VideoCapture is live camera.
        None => Grabber::new(
            config.index.into(),
            config.height.into(),
            config.width.into(),
            config.framerate.into(),
        ),
    }
    .unwrap();

    // Instance of the motion detector.
    let mut detector = MotionDetector::new();

    // Instance of the frame writer.
    let mut writer = Writer::new(
        &filename,
        Codec::XVID,
        grabber.get_fps(),
        grabber.get_size(),
        config.overlay,
    )
    .unwrap();

    // Print config options if config.quiet is false.
    if !config.quiet {
        println!("{:#?}", &config);
    }

    // Acquire N frames.
    for _ in 0..N {
        frames.push(grabber.grab().unwrap());
    }

    // Save the start time.
    let start = Instant::now();
    for frame in frames {
        match detector.detect_motion(frame) {
            Ok(frame) => {
                if let Some(frame) = frame {
                    // If frame is detected, write it to the file.
                    writer.write(frame).unwrap();
                    // Count the detected frames.
                    detected_frames += 1;
                }
            }
            Err(_) => panic!("not enaugh frames to run the test!"),
        }
    }

    //Calculate the elapsed time to process motion detection on all the N frmaes.
    let tot_dur_ns = start.elapsed();
    let dur_ns = tot_dur_ns.div_f32(N as f32);
    let max = 1e3 / config.framerate as f32;
    println!("==> # saved frames: {}", detected_frames);
    println!("==> processing motion detection took: {:?}", tot_dur_ns);
    println!(
        "==> processing motion detection per frame took (avg): {:?}",
        dur_ns
    );
    println!("==> max value allowed: {}ms", max);

    // Remove generated video file.
    fs::remove_file(filename).expect("unable to remove output file.");

    assert!(dur_ns.subsec_micros() <= (max * 1e3) as u32);
}


src/args.rs
// bombuscv: OpenCV based motion detection/recording software built for research on bumblebees.
// Copyright (C) 2022 Marco Radocchia
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see https://www.gnu.org/licenses/.

use crate::{config::expand_home, Codec};
use clap::ArgAction::{Set, SetTrue};
pub use clap::Parser;
use std::{fs, path::PathBuf};

/// Custom parser for `directory` field.
/// Automatically expands ~ and creates directory if doesn't exist.
pub fn parse_directory(directory: &str) -> Result<PathBuf, String> {
    let path = expand_home(&PathBuf::from(directory));
    if !path.is_dir() && fs::create_dir_all(&path).is_err() {
        return Err(String::from("unable to create specified directory"));
    }

    Ok(path)
}

/// Custom parser for `codec` field.
pub fn parse_codec(codec: &str) -> Result<Codec, String> {
    Ok(match codec {
        "h264" => Codec::H264,
        "mjpg" => Codec::MJPG,
        "xvid" => Codec::XVID,
        "mp4v" => Codec::MP4V,
        _ => return Err("unsupported codec".to_string()),
    })
}

/// Custom parser for `video` field.
fn parse_video(video: &str) -> Result<PathBuf, String> {
    let video = expand_home(&PathBuf::from(video));
    match video.is_file() {
        true => Ok(video),
        false => Err(String::from("the given path is not a file")),
    }
}

/// OpenCV motion detection/video-recording tool developed for research on Bumblebees.
#[derive(Parser, Debug)]
#[clap(
    author = "Marco Radocchia <marco.radocchia@outlook.com>",
    version,
    about,
    long_about = None
)]
pub struct Args {
    /// /dev/video<INDEX> capture camera index.
    #[clap(short, long, action = Set)]
    pub index: Option<u8>,

    /// Video file as input.
    #[clap(
        short,
        long,
        value_parser = parse_video,
        conflicts_with_all = &["index", "overlay", "height", "width", "framerate"]
    )]
    pub video: Option<PathBuf>,

    /// Video capture frame height.
    #[clap(short = 'H', long, action = Set)]
    pub height: Option<u16>,

    /// Video capture frame width.
    #[clap(short = 'W', long, action = Set)]
    pub width: Option<u16>,

    /// Video capture framerate.
    #[clap(short, long, action = Set)]
    pub framerate: Option<u8>,

    /// Video codec.
    #[clap(
        short,
        long,
        possible_values = ["h264", "mjpg", "xvid", "mp4v"],
        value_parser = parse_codec
    )]
    pub codec: Option<Codec>,

    /// Output video directory.
    #[clap(short, long, value_parser = parse_directory)]
    pub directory: Option<PathBuf>,

    /// Output video filename format (see
    /// <https://docs.rs/chrono/latest/chrono/format/strftime/index.html> for valid specifiers).
    #[clap(long, action = Set)]
    pub format: Option<String>,

    /// Date&Time video overlay.
    #[clap(short, long, action = SetTrue)]
    pub overlay: bool,

    /// Disable colored output.
    #[clap(long, action = SetTrue)]
    pub no_color: bool,

    /// Mute standard output.
    #[clap(short, long, action = SetTrue)]
    pub quiet: bool,
}


src/main.rs
// bombuscv: OpenCV based motion detection/recording software built for research on bumblebees.
// Copyright (C) 2022 Marco Radocchia
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see https://www.gnu.org/licenses/.

#[cfg(test)]
mod test;

use bombuscv_rs::{
    args::{Args, Parser},
    color::{Colorizer, MsgType},
    config::Config,
    Codec, Grabber, MotionDetector, Writer,
};
use chrono::Local;
use signal_hook::{consts::SIGINT, flag::register};
use std::io;
use std::{
    path::Path,
    process,
    sync::{
        atomic::{AtomicBool, Ordering},
        mpsc, Arc,
    },
    thread,
};

fn main() -> io::Result<()> {
    // Parse CLI arguments.
    let args = Args::parse();
    // Parse config file and override options with CLI arguments.
    let config = match Config::parse() {
        Ok(config) => config,
        Err(e) => {
            Colorizer::new(MsgType::Error, args.no_color, "error [config]", e).print()?;
            Colorizer::new(
                MsgType::Warn,
                args.no_color,
                "warning",
                "using default configuration",
            )
            .print()?;
            Config::default()
        }
    }
    .override_with_args(args);

    // Format video file path as <config.directory/date&time>.
    let filename = Local::now()
        .format(
            config
                .directory
                // Output video file name (derived by file format) + extension.
                .join(Path::new(&config.format).with_extension("mkv"))
                // Convert Path object to string.
                .to_str()
                .unwrap(),
        )
        .to_string();

    // Instance of the frame grabber.
    let grabber = match &config.video {
        // VideoCapture is video file.
        Some(video) => Grabber::from_file(video),
        // VideoCapture is live camera.
        None => Grabber::new(
            config.index.into(),
            config.height.into(),
            config.width.into(),
            config.framerate.into(),
        ),
    };
    let grabber = match grabber {
        Ok(grabber) => grabber,
        Err(e) => {
            Colorizer::new(MsgType::Error, config.no_color, "error", e).print()?;
            process::exit(1);
        }
    };

    // Print info.
    if !config.quiet {
        let mut colorizer = Colorizer::empty(MsgType::Info, config.no_color);

        let input = if let Some(video) = &config.video {
            video.display().to_string()
        } else {
            format!("/dev/video{}", &config.index)
        };

        let messages = vec![
            ("==> Input", input),
            ("==> Framerate", grabber.get_fps().to_string()),
            (
                "==> Frame size",
                format!("{}x{}", grabber.get_width(), grabber.get_height()),
            ),
            ("==> Printing overlay", format!("{}", config.overlay)),
            ("==> Output video file", filename.clone()),
        ];

        for msg in messages {
            colorizer.update(msg.0, msg.1);
            colorizer.print()?;
        }
    }

    // Instance of the motion detector.
    let detector = MotionDetector::new();

    // Instance of the frame writer.
    let writer = match Writer::new(
        &filename,
        Codec::XVID,
        grabber.get_fps(),
        grabber.get_size(),
        config.overlay,
        config.overlay_border,
    ) {
        Ok(writer) => writer,
        Err(e) => {
            Colorizer::new(MsgType::Error, config.no_color, "error", e).print()?;
            process::exit(1);
        }
    };

    // Save memory dropping `filename`.
    drop(filename);

    // Run the program.
    run(grabber, detector, writer, config.no_color)?;

    // Gracefully terminated execution.
    if !config.quiet {
        Colorizer::new(MsgType::Info, config.no_color, "\nbombuscv", "done!").print()?;
    }

    Ok(())
}

/// Run `bombuscv`: spawn & join frame grabber, detector and writer threads.
fn run(
    mut grabber: Grabber,
    mut detector: MotionDetector,
    mut writer: Writer,
    no_color: bool,
) -> io::Result<()> {
    // Create channels for message passing between threads.
    // NOTE: using mpsc::sync_channel (blocking) to avoid channel size
    // growing indefinitely, resulting in infinite memory usage.
    let (raw_tx, raw_rx) = mpsc::sync_channel(100);
    let (proc_tx, proc_rx) = mpsc::sync_channel(100);

    // Spawn frame grabber thread:
    // this thread captures frames and passes them to the motion detecting thread.
    let grabber_handle = thread::spawn(move || -> io::Result<()> {
        let term = Arc::new(AtomicBool::new(false));
        // Register signal hook for SIGINT events: in this case error is unrecoverable, so report
        // it to the user & exit process with code error code.
        if let Err(e) = register(SIGINT, Arc::clone(&term)) {
            Colorizer::new(
                MsgType::Error,
                no_color,
                "fatal error",
                format!("unable to register SIGINT hook '{e}'"),
            )
            .print()?;
            process::exit(1);
        };

        // Start grabber loop: loop guard is 'received SIGINT'.
        while !term.load(Ordering::Relaxed) {
            let frame = match grabber.grab() {
                Ok(frame) => frame,
                Err(e) => {
                    Colorizer::new(MsgType::Warn, no_color, "warning", e).print()?;
                    continue;
                }
            };

            // Grab frame and send it to the motion detection thread.
            if raw_tx.send(frame).is_err() {
                break;
            }
        }

        Ok(())
    });

    // Spawn motion detection thread:
    // this thread receives frames from the grabber thread, processes it and if motion is detected,
    // passes the frame to the frame writing thread.
    let detector_handle = thread::spawn(move || -> io::Result<()> {
        // Loop over received frames from the frame grabber.
        for frame in raw_rx {
            match detector.detect_motion(frame) {
                // Valid frame is received.
                Ok(val) => {
                    // Motion has been detected: send frame to the video writer.
                    if let Some(frame) = val {
                        if proc_tx.send(frame).is_err() {
                            Colorizer::new(
                                MsgType::Warn,
                                no_color,
                                "warning",
                                "unable to send processed frame to video output",
                            )
                            .print()?;
                        };
                    }
                }
                // Last captured frame was an empty frame: no more input is provided, interrupt the
                // thread (break the loop).
                Err(_) => break,
            }
        }

        Ok(())
    });

    // Spawn frame writer thread:
    // this thread receives the processed frames by the motion detecting thread and writes them in
    // the output video output.
    let writer_handle = thread::spawn(move || -> io::Result<()> {
        // Loop over received frames from the motion detector.
        for frame in proc_rx {
            // Write processed frames (motion detected) to the video output.
            if let Err(e) = writer.write(frame) {
                Colorizer::new(MsgType::Warn, no_color, "warning", e).print()?;
            };
        }

        Ok(())
    });

    // Join all threads.
    grabber_handle.join().expect("cannot join grabber thread")?;
    detector_handle
        .join()
        .expect("cannot join detector thread")?;
    writer_handle.join().expect("cannot join writer thread")?;

    Ok(())
}


src/config.rs
// bombuscv: OpenCV based motion detection/recording software built for research on bumblebees.
// Copyright (C) 2022 Marco Radocchia
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see https://www.gnu.org/licenses/.

use crate::{args::Args, error::ErrorKind, Codec};
use directories::BaseDirs;
use serde::{de, Deserialize, Deserializer};
use std::{
    env,
    fmt::Debug,
    fs,
    path::{Path, PathBuf},
    string::String,
};

/// Expands `~` in `path` to absolute HOME path.
pub fn expand_home(path: &Path) -> PathBuf {
    if let Ok(path) = path.strip_prefix("~") {
        // Generate the absolute path for HOME.
        BaseDirs::new()
            .expect("unable to find home directory")
            .home_dir()
            .to_path_buf()
            .join(path)
    } else {
        path.to_path_buf()
    }
}

/// Custom deserializer for `directory` field.
/// Automatically expands ~ and creates directory if doesn't exist.
fn deserialize_directory<'de, D>(directory: D) -> Result<PathBuf, D::Error>
where
    D: Deserializer<'de>,
{
    let path = expand_home(&PathBuf::deserialize(directory)?);
    if !path.is_dir() && fs::create_dir_all(&path).is_err() {
        return Err(de::Error::custom("unable to create specified directory"));
    }

    Ok(path)
}

/// Custom deserializer for `codec` field.
fn deserialize_codec<'de, D>(codec: D) -> Result<Codec, D::Error>
where
    D: Deserializer<'de>,
{
    let codec = String::deserialize(codec)?;
    Ok(match codec.as_str() {
        "h264" => Codec::H264,
        "mjpg" => Codec::MJPG,
        "xvid" => Codec::XVID,
        "mp4v" => Codec::MP4V,
        _ => return Err(de::Error::custom("unsupported codec")),
    })
}

/// Default value for /dev/video<index> capture camera index.
fn default_index() -> u8 {
    0
}

/// Default value for video capture frame height.
fn default_height() -> u16 {
    480
}

/// Default value for video capture frame width.
fn default_width() -> u16 {
    640
}

/// Default value for video capture framerate.
fn default_framerate() -> u8 {
    60
}

/// Default video codec.
fn default_codec() -> Codec {
    Codec::H264
}

/// Default output video directory.
fn default_directory() -> PathBuf {
    // No base directories could be determined, so panicking is fine here.
    BaseDirs::new()
        .expect("unable to find HOME directory")
        .home_dir()
        .to_path_buf()
}

/// Default output video filename format, for example: `2022-06-23T11:49:00`.
fn default_format() -> String {
    String::from("%Y-%m-%dT%H:%M:%S")
}

/// Default value for Date&Time overlay border.
fn default_overlay_border() -> u8 {
    3
}

/// Configuration options.
#[derive(Deserialize, Debug)]
pub struct Config {
    /// /dev/video<index> capture camera index.
    #[serde(default = "default_index")]
    pub index: u8,

    /// Video file as input.
    #[serde(skip_deserializing)]
    pub video: Option<PathBuf>,

    /// Video capture frame height.
    #[serde(default = "default_height")]
    pub height: u16,

    /// Video capture frame width.
    #[serde(default = "default_width")]
    pub width: u16,

    /// Video capture framerate.
    #[serde(default = "default_framerate")]
    pub framerate: u8,

    /// Video codec.
    #[serde(default = "default_codec", deserialize_with = "deserialize_codec")]
    pub codec: Codec,

    /// Output video directory.
    #[serde(
        default = "default_directory",
        deserialize_with = "deserialize_directory"
    )]
    pub directory: PathBuf,

    /// Output video filename format (see
    /// <https://docs.rs/chrono/latest/chrono/format/strftime/index.html> for valid specifiers).
    #[serde(default = "default_format")]
    pub format: String,

    /// Date&Time video overlay.
    #[serde(default)]
    pub overlay: bool,

    /// Date&Time video overlay border.
    #[serde(default = "default_overlay_border")]
    pub overlay_border: u8,

    /// Disable colored output.
    #[serde(skip_deserializing, default)]
    pub no_color: bool,

    /// Mute standard output.
    #[serde(default)]
    pub quiet: bool,
}

/// Implement the Default trait for Config.
impl Default for Config {
    /// Default configuration.
    fn default() -> Self {
        Self {
            index: default_index(),
            video: None,
            height: default_height(),
            width: default_width(),
            framerate: default_framerate(),
            codec: default_codec(),
            directory: default_directory(),
            format: default_format(),
            overlay: false,
            overlay_border: default_overlay_border(),
            no_color: false,
            quiet: false,
        }
    }
}

impl Config {
    /// Parse configuration from config file, return Err on error.
    pub fn parse() -> Result<Self, ErrorKind> {
        if let Some(base_dirs) = BaseDirs::new() {
            // Fetch the environment variables for BOMBUSCV_CONFIG to hold a custom path to store
            // the configuration file.
            let config_file = if let Ok(path) = env::var("BOMBUSCV_CONFIG") {
                expand_home(Path::new(&path))
            } else {
                // XDG base spec directories: ~/.config/bomsucv/config.toml
                base_dirs
                    .config_dir()
                    .join(Path::new("bombuscv/config.toml"))
            };

            // On Err variant, return empty string (=> default config).
            let config = fs::read_to_string(config_file).unwrap_or_default();

            // Deserialize toml configuration file into Config.
            toml::from_str(&config).map_err(|e| ErrorKind::BrokenConfig(e.to_string()))
        } else {
            Err(ErrorKind::ConfigNotFound)
        }
    }

    /// Override configuration with command line arguments.
    pub fn override_with_args(mut self, args: Args) -> Self {
        if let Some(directory) = args.directory {
            self.directory = directory;
        }

        if let Some(format) = args.format {
            self.format = format;
        }

        if args.no_color {
            self.no_color = true;
        }

        if args.quiet {
            self.quiet = true;
        }

        // Input is video: disable overlay.
        if let Some(video) = args.video {
            self.video = Some(video);
            if self.overlay {
                self.overlay = false;
            }
        }

        if let Some(index) = args.index {
            self.index = index;
        }

        if let Some(height) = args.height {
            self.height = height;
        }

        if let Some(width) = args.width {
            self.width = width;
        }

        if let Some(framerate) = args.framerate {
            self.framerate = framerate;
        }

        if let Some(codec) = args.codec {
            self.codec = codec;
        }

        if args.overlay {
            self.overlay = true;
        }

        self
    }
}
